src/HaroohiePals.NitroKart/MapObj/Obstacles/KoopaBlock.cs

```diff
@@ -0,0 +1,151 @@
+            Flags &= ~InstanceFlags.Bit3;
+            Alpha = 31;
+            _stateMachine.NextState = KoopaBlockState.SpeedUp;
+            _stateMachine.GotoNextState = true;
+        }
+
+        private void SpeedUpStateInit()
+        {
+        }
+
+        private void SpeedUpStateUpdate()
+        {
+            double sin;
+            double val;
+            double val2;
+
+            sin = MObjUtil.SinIdx((ushort)(_pathwalker.Progress * 64.0));
+            val = sin * _speed;
+            _pathwalker.Speed = val > 0.125 ? val : 0.125;
+            val2 = _pathwalker.Path.Parts[_pathwalker.PartIdx].OneDivLength;
+            if (val <= 0.125)
+                val = 0.125;
+            _pathwalker.PartSpeed = val * val2;
+            if (_pathwalker.Progress < 0.0625)
+                return;
+            _stateMachine.NextState = KoopaBlockState.Move;
+            _stateMachine.GotoNextState = true;
+        }
+
+        private void MoveStateinit()
+        {
+            double speed = _speed;
+            _pathwalker.Speed = speed;
+            _pathwalker.PartSpeed = speed * _pathwalker.Path.Parts[_pathwalker.PartIdx].OneDivLength;
+        }
+
+        private void MoveStateUpdate()
+        {
+            if (_pathwalker.CurPoit.Duration == 0 && _pathwalker.HasEnded == false)
+                return;
+            if (_pathwalker.Progress < 0.9375)
+                return;
+            _stateMachine.NextState = KoopaBlockState.SlowDown;
+            _stateMachine.GotoNextState = true;
+        }
+
+        private void SlowDownStateInit()
+        {
+        }
+
+        private void SlowDownStateUpdate()
+        {
+            double sin;
+            double val;
+
+            sin = -(MObjUtil.SinIdx((ushort)(_pathwalker.Progress * 64.0)));
+            val = sin * _speed;
+            _pathwalker.Speed = val > 0.125 ? val : 0.125;
+            if (val <= 0.125)
+                val = 0.125;
+            _pathwalker.PartSpeed = val * _pathwalker.Path.Parts[_pathwalker.PartIdx].OneDivLength;
+        }
+
+        public virtual void Update()
+        {
+            if (_waitCounter != 0)
+            {
+                _waitCounter -= 1;
+                Velocity = Vector3d.Zero;
+            }
+            else
+            {
+                if (_pathwalker.Update())
+                {
+                    if (_stateMachine.CurState == KoopaBlockState.SlowDown)
+                    {
+                        _stateMachine.NextState = KoopaBlockState.SpeedUp;
+                        _stateMachine.GotoNextState = true;
+                    }
+                    _waitCounter = (ushort)_pathwalker.PrevPoit.Duration;
+                }
+                _lastPosition = Position;
+                Position = _pathwalker.CalcCurrentPointLinearXYZ();
+                _stateMachine.Execute();
+                Velocity = Position - _lastPosition;
+                _basePos = (-_size.Y * Mtx.Row1) + Position;
```

Comment:
```suggestion
                _basePos = Position -_size.Y * Mtx.Row1;
```
I think this should be equivalent and be more clear. We're subtracting the height of the block, along the Y axis